[강진호] week9

jinhok96 opened the pull request:
## 요구사항

### 기본

- [x] 링크를 입력하고 “추가하기” 버튼을 누르면 “폴더에 추가” 모달이 보이나요?
- [x] “폴더 추가” 버튼을 누르면 “폴더 추가” 모달이 보이나요?
- [x] “공유” 버튼을 누르면 “폴더 공유” 모달이 보이나요?
- [x] “폴더 추가” 버튼을 누르면 “폴더 추가” 모달이 보이나요?
- [x] “삭제” 버튼을 누르면 “폴더 삭제” 모달이 보이나요?
- [x] 케밥 버튼을 누르면 “삭제하기”, “폴더에 추가” 버튼이 있는 팝오버가 보이나요?
- [x] 팝오버에 있는 “삭제하기” 버튼을 누르면 “링크 삭제” 모달이 보이나요?
- [x] 팝오버에 있는 “폴더에 추가” 버튼을 누르면 “폴더에 추가” 모달이 보이나요?
- [x] “폴더 공유” 모달에서 “카카오톡 아이콘”을 클릭하면 카카오로 공유 폴더 페이지 링크 공유하기 가능한가요?
- [x] “폴더 공유” 모달에서 “페이스북 아이콘”을 클릭하면 페이스북으로 공유 폴더 페이지 링크 공유하기 가능한가요?
- [x] “폴더 공유” 모달에서 “링크 아이콘”을 클릭하면 클립보드에 공유 폴더 페이지 링크가 복사 되나요?

## 주요 변경사항

- 케밥 버튼 클릭 시 나타나는 팝오버 메뉴는 카드 컴포넌트에서 마우스가 벗어나면 사라집니다.
- 모든 모달은 우측 상단 X 버튼을 누르거나, 모달 외부를 클릭하거나, Esc를 누르면 사라집니다.
- 카카오톡 공유는 api를 획득하기 어려운 문제로 인해 링크 복사 로직으로 대체하였습니다. 
- 홈페이지에서 버튼을 통해 각 페이지로 이동 가능합니다.

## 스크린샷

![image](이미지url)

## 멘토에게

-
-
- 셀프 코드 리뷰를 통해 질문 이어가겠습니다.

jinhok96 committed:
feat: SortingButton에 hover 스타일링 추가
	modified:   src/assets/styles/colors.css
	modified:   src/components/Common/Button.js
	modified:   src/components/Common/SortingButton.js
	modified:   src/components/Main/SortingSection.js

jinhok96 committed:
chore: 여러 modal 파일 생성
	new file:   src/components/Modals/AddFolderModal.js
	new file:   src/components/Modals/AddFolderModal.module.css
	new file:   src/components/Modals/AddModal.js
	new file:   src/components/Modals/AddModal.module.css
	new file:   src/components/Modals/DeleteFolderModal.js
	new file:   src/components/Modals/DeleteFolderModal.module.css
	new file:   src/components/Modals/DeleteLinkModal.js
	new file:   src/components/Modals/DeleteLinkModal.module.css
	new file:   src/components/Modals/EditModal.js
	new file:   src/components/Modals/EditModal.module.css
	new file:   src/components/Modals/ShareModal.js
	new file:   src/components/Modals/ShareModal.module.css

jinhok96 committed:
test: AddFolderModal 테스트
	new file:   src/assets/images/close.svg
	modified:   src/assets/styles/colors.css
	modified:   src/components/Common/AddLInkButton.js
	modified:   src/components/Common/LoginButton.js
	new file:   src/components/Common/ModalButton.js
	new file:   src/components/Common/ModalButton.module.css
	new file:   src/components/Common/TextInput.js
	new file:   src/components/Common/TextInput.module.css
	modified:   src/components/Main/SearchBar.js
	new file:   src/components/Modal/AddFolderModal.js
	new file:   src/components/Modal/AddFolderModal.module.css
	renamed:    src/components/Modals/AddModal.js -> src/components/Modal/AddModal.js
	renamed:    src/components/Modals/AddModal.module.css -> src/components/Modal/AddModal.module.css
	renamed:    src/components/Modals/DeleteFolderModal.js -> src/components/Modal/DeleteFolderModal.js
	renamed:    src/components/Modals/DeleteFolderModal.module.css -> src/components/Modal/DeleteFolderModal.module.css
	renamed:    src/components/Modals/DeleteLinkModal.js -> src/components/Modal/DeleteLinkModal.js
	renamed:    src/components/Modals/DeleteLinkModal.module.css -> src/components/Modal/DeleteLinkModal.module.css
	renamed:    src/components/Modals/EditModal.js -> src/components/Modal/EditModal.js
	renamed:    src/components/Modals/EditModal.module.css -> src/components/Modal/EditModal.module.css
	new file:   src/components/Modal/ModalContainer.js
	new file:   src/components/Modal/ModalContainer.module.css
	renamed:    src/components/Modals/ShareModal.js -> src/components/Modal/ShareModal.js
	renamed:    src/components/Modals/ShareModal.module.css -> src/components/Modal/ShareModal.module.css
	deleted:    src/components/Modals/AddFolderModal.js
	deleted:    src/components/Modals/AddFolderModal.module.css
	modified:   src/pages/HomePage/HomePage.js

jinhok96 committed:
chore: 컴포넌트 이름 변경
	renamed:    src/components/Modal/AddModal.js -> src/components/Modal/AddToFolderModal.js
	renamed:    src/components/Modal/AddModal.module.css -> src/components/Modal/AddToFolderModal.module.css
	renamed:    src/components/Modal/EditModal.js -> src/components/Modal/EditFolderNameModal.js
	renamed:    src/components/Modal/EditModal.module.css -> src/components/Modal/EditFolderNameModal.module.css

jinhok96 committed:
test: EditFolderNameModal 테스트
	modified:   src/components/Modal/EditFolderNameModal.js
	modified:   src/components/Modal/EditFolderNameModal.module.css
	modified:   src/pages/HomePage/HomePage.js

jinhok96 committed:
test: ShareModal 테스트
	new file:   src/assets/images/share-facebook.svg
	new file:   src/assets/images/share-kakao.svg
	new file:   src/assets/images/share-link-copy.svg
	modified:   src/components/Modal/ShareModal.js
	modified:   src/components/Modal/ShareModal.module.css
	modified:   src/pages/HomePage/HomePage.js
	new file:   src/utils/getHostAddress.js

jinhok96 committed:
chore: 잘못된 z-index 수정
	modified:   src/assets/styles/utility.css
	modified:   src/components/Main/Card.js
	modified:   src/components/Main/Card.module.css
	modified:   src/components/Main/CardList.js
	modified:   src/components/Modal/ModalContainer.js

jinhok96 committed:
chore: 카카오톡 공유 관련 주석 추가
	modified:   src/components/Modal/ShareModal.js

jinhok96 committed:
chore: 굳이 필요하지 않은 함수 제거
	modified:   src/components/Modal/ShareModal.js
	deleted:    src/utils/getHostAddress.js

jinhok96 committed:
chore: 디버그용 콘솔 출력문 제거
	modified:   src/components/Modal/ShareModal.js

jinhok96 committed:
test: DeleteFolderModal 테스트
	modified:   src/assets/styles/colors.css
	modified:   src/components/Modal/AddFolderModal.js
	modified:   src/components/Modal/DeleteFolderModal.js
	modified:   src/components/Modal/DeleteFolderModal.module.css
	modified:   src/components/Modal/EditFolderNameModal.js
	modified:   src/components/Modal/ShareModal.js
	modified:   src/pages/HomePage/HomePage.js

jinhok96 committed:
test: DeleteLinkModal 테스트
	modified:   src/components/Modal/DeleteLinkModal.js
	modified:   src/components/Modal/DeleteLinkModal.module.css
	modified:   src/pages/HomePage/HomePage.js

jinhok96 committed:
test: AddToFolderModal 테스트
	new file:   src/assets/images/check.svg
	modified:   src/assets/styles/utility.css
	new file:   src/components/Common/AddFolderList.js
	new file:   src/components/Common/AddFolderList.module.css
	modified:   src/components/Main/CardList.js
	modified:   src/components/Modal/AddToFolderModal.js
	modified:   src/components/Modal/AddToFolderModal.module.css
	modified:   src/components/Modal/DeleteFolderModal.js
	modified:   src/components/Modal/DeleteLinkModal.js
	modified:   src/components/Modal/DeleteLinkModal.module.css
	modified:   src/pages/HomePage/HomePage.js

jinhok96 committed:
test: SelectMenu 테스트
	new file:   src/components/Common/SelectMenu.js
	new file:   src/components/Common/SelectMenu.module.css
	modified:   src/components/Modal/AddToFolderModal.js
	modified:   src/pages/HomePage/HomePage.js

jinhok96 committed:
chore: select, hover 상태에 따른 스타일 관리 로직 개선
	modified:   src/assets/styles/colors.css
	modified:   src/components/Common/SelectMenu.js
	modified:   src/components/Common/SortingButton.js
	modified:   src/components/Main/SortingSection.js
	modified:   src/components/Modal/AddToFolderModal.js
	modified:   src/pages/HomePage/HomePage.js

jinhok96 committed:
chore: Card에 SelectMenu 팝오버 추가
	modified:   src/components/Common/SelectMenu.js
	modified:   src/components/Common/SelectMenu.module.css
	modified:   src/components/Main/Card.js
	modified:   src/components/Main/Card.module.css

jinhok96 committed:
feat: SelectMenu에 모달 연결
- 각 메뉴 클릭하면 해당 모달 렌더링
- closeIcon 또는 Esc로 모달 닫기
	modified:   src/components/Common/SelectMenu.js
	modified:   src/components/Main/Card.js
	modified:   src/components/Modal/AddToFolderModal.js
	modified:   src/components/Modal/DeleteLinkModal.js
	modified:   src/components/Modal/ModalContainer.js
	modified:   src/components/Modal/ModalContainer.module.css

jinhok96 committed:
chore: 모든 modalContainer 컴포넌트에 props 추가
	modified:   src/components/Modal/AddFolderModal.js
	modified:   src/components/Modal/DeleteFolderModal.js
	modified:   src/components/Modal/EditFolderNameModal.js
	modified:   src/components/Modal/ShareModal.js

jinhok96 committed:
feat: 모달에 link 연결
	modified:   src/components/Common/SelectMenu.js
	modified:   src/components/Main/Card.js
	modified:   src/components/Modal/AddToFolderModal.js

jinhok96 committed:
chore: 필요없는 PropTypes 제거
	modified:   src/components/Modal/DeleteFolderModal.js
	modified:   src/components/Modal/DeleteLinkModal.js

jinhok96 committed:
chore: AddLInkBar에 값이 있을 때 버튼 누르면 모달 렌더링
	modified:   src/components/Common/AddLInkButton.js
	modified:   src/components/Common/SelectMenu.js
	modified:   src/components/Header/AddLinkBar.js
	modified:   src/components/Header/AddLinkBar.module.css

jinhok96 committed:
feat: 폴더 추가 버튼 클릭하면 모달 렌더링
	modified:   src/components/Main/SortingSection.js

jinhok96 committed:
feat: 마우스 클릭이 가능한 요소에 cursor: pointer 속성 추가
	modified:   src/assets/styles/utility.css
	modified:   src/components/Common/Button.js
	modified:   src/components/Modal/AddToFolderModal.js

jinhok96 committed:
feat: Option 컴포넌트에 cursor: pointer 추가
	modified:   src/components/Common/Option.js

jinhok96 committed:
chore: Option 컴포넌트에 Button 컴포넌트 추가
	modified:   src/components/Common/Option.js

jinhok96 committed:
feat: Option 클릭 시 모달 렌더링
	modified:   src/components/Main/SortingSection.js
	modified:   src/components/Modal/EditFolderNameModal.js

jinhok96 committed:
fix: 잘못된 folder 객체 속성 변경
	modified:   src/components/Modal/ShareModal.js

jinhok96 committed:
feat: 홈페이지에 각 페이지로 연결되는 버튼 추가
	modified:   src/pages/HomePage/HomePage.js
	modified:   src/pages/HomePage/HomePage.module.css

jinhok96 committed:
fix: 잘못된 주소 수정
	modified:   src/pages/HomePage/HomePage.js

SeolJaeHyeok commented on src/components/Common/SelectMenu.js:
```diff
@@ -0,0 +1,95 @@
+        break;
+      default:
+        break;
+    }
```
`1`, `2` 보다 모달을 특정할 수 있는 구체적인 이름을 설정하는게 좋아보여요!

SeolJaeHyeok commented on src/components/Common/SelectMenu.js:
```diff
@@ -0,0 +1,95 @@
+  };
+
+  const handleMenuClick = (e, key) => {
+    e.stopPropagation();
+    const targetMenu = menuList.find((menu) => menu.id === key);
+    setSelectedMenu(targetMenu);
+    setIsModalOpen(true);
```
`e.stopPropagation()`을 하신 이유가 있을까요?

jinhok96 replied:
이벤트 버블링때문에 클릭 이벤트가 부모 요소인 Card에 전파되어 Card의 onClick가 작동하는 문제가 있어서 추가했었는데 ReactModal을 사용하면서 제거했습니다!

SeolJaeHyeok commented on src/components/Common/SelectMenu.js:
```diff
@@ -0,0 +1,95 @@
+            ${hoveredMenu?.id === menu.id ? hoveredMenuClasses : defaultMenuClasses}`}
+            text={menu.label}
+            onClick={(e) => handleMenuClick(e, menu.id)}
+            onMouseEnter={() => handleMenuMouseEnter(menu.id)}
+            onMouseLeave={handleMenuMouseLeave}
```
마우스 포인터가 해당 버튼 호버링될 때마다 `setHoveredMenu`을 시켜줘서 불필요한 렌더링이 일어날 거 같은데 이 부분을 다른 이벤트로 개선해보면 좋을거 같아요!

SeolJaeHyeok commented on src/components/Common/TextInput.js:
```diff
@@ -0,0 +1,33 @@
+
+import styles from 'components/Common/TextInput.module.css';
+
+function TextInput({ className, value, onChange, placeholder }) {
+  const inputClasses = classNames(
+    styles.input,
+    'background-white',
+    'text-color-gray100',
+    'border-gray20-1px',
+    className
+  );
+
+  return <input className={inputClasses} type="text" value={value} onChange={onChange} placeholder={placeholder} />;
+}
```
필요한 값은 사용하는 쪽에서 주입받아서 하는 방법으로 구현하셨네요👍 👍

SeolJaeHyeok commented on src/components/Header/AddLinkBar.js:
```diff
@@ -5,26 +5,31 @@ import addLinkIcon from 'assets/images/link.svg';
     setAddLinkValue(e.target.value);
   };
 
+  const openModal = () => {
+    const link = { url: addLinkValue };
+    return <AddToFolderModal link={link} isModalOpen={isModalOpen} setIsModalOpen={setIsModalOpen} />;
+  };
```
`setIsModalOpen`이라는 setState 함수를 넘기는게 아닌 setState를 바꿔주는 콜백 함수를 넘겨 외부에서 의존성을 주입하고 컴포넌트 내부는 주입받은 의존성을 실행하는 용도로만 동작시키는게 좋을거 같아요!

SeolJaeHyeok commented on src/components/Header/AddLinkBar.js:
```diff
@@ -35,10 +40,9 @@ function AddLinkBar() {
-        <AddLinkButton className={addLinkButtonClasses} />
-      </div>
+      <img className={inputImgClasses} src={addLinkIcon} alt="addLinkIcon" />
+      <AddLinkButton className={addLinkButtonClasses} onClick={handleButtonClick} />
+      {isModalOpen && openModal()}
```
멘토링 때 말씀드렸지만 modal의 경우 일반적으로 컴포넌트마다 렌더링 시키기보다는 전역에서 렌더링 시키는게 관리, 유지, 보수, 사용성 측면에서 좋은 방법이에요!

전역 상태 라이브러리를 사용해도 좋지만 Context API를 사용해서 Modal을 관리하는 방법도 있어요! 
관련 링크 첨부해드릴테니 가볍게 한 번 봐주세요!

[참고](https://nakta.dev/how-to-manage-modals-1)

jinhok96 replied:
첨부해주신 링크 참고해서 모달 구현 로직 수정했습니다. Context API, ReactModal, loadable, useMemo 추가했고 모달을 중앙에서 한 번에 관리하도록 구현했습니다.

로직은 전부는 아니지만 어느 정도 필요한 디버깅은 할 수 있는 정도로 이해하고 있습니다. 콘솔창에 'React-Modal: Cannot register modal instance that's already open' 오류가 뜨는데 확인한 결과 StrictMode에 의해 발생하는 것이어서 무시했습니다. 

모달 구현 관련해서 리뷰 부탁드립니다!

SeolJaeHyeok commented on src/components/Main/SortingSection.js:
```diff
@@ -12,13 +13,22 @@ import ErrorMessage from 'components/Common/ErrorMessage';
   const ALL = { id: 0, name: '전체' };
 
+  const [hoveredFolder, setHoveredFolder] = useState(null);
+  const [isAddFolderModalOpen, setIsAddFolderModalOpen] = useState(false);
+  const [isOptionListModalOpen, setIsOptionListModalOpen] = useState(false);
```
각 컴포넌트마다 모달 상태를 관리하게 되면 지금처럼 모달의 종류마다 상태를 가지게 되는 문제가 발생할 수 있으니 전역으로 모달을 처리하는 방법을 고민해보도록 합시다!

그렇지 않고 만약 컴포넌트마다 관리를 하게 된다면 지금처럼 `isAddFolderModalOpen`, `isOptionsListModalOpen`의 상태를 따로 두지 않고 아래와 같이 모든 modal 상태를 관리하는 상태를 만드는게 관심사를 응집하는 방법 중 하나입니다!
```suggestion
  const [modalOpenState, setModalOpenState] = useState({
    isAddFolderOpen: false,
    isOptionsListOpne: false,
  })
  
  ...
  const handleAddFolderButtonClick = () => {
    setModalOpenState(prev => ({...prev, isAddFolderOpen: true})
  };
  
```

SeolJaeHyeok commented on src/components/Main/SortingSection.js:
```diff
@@ -37,24 +47,86 @@ function SortingSection({ selectedFolder, setSelectedFolder }) {
+    return <AddFolderModal isModalOpen={isAddFolderModalOpen} setIsModalOpen={setIsAddFolderModalOpen} />;
+  };
+
+  const openOptionListModal = () => {
+    let modal;
+    switch (selectedOptionKey) {
+      case 1:
+        modal = (
+          <ShareModal
+            folder={selectedFolder}
+            isModalOpen={isOptionListModalOpen}
+            setIsModalOpen={setIsOptionListModalOpen}
+          />
+        );
+        break;
+      case 2:
+        modal = (
+          <EditFolderNameModal
+            folder={selectedFolder}
+            isModalOpen={isOptionListModalOpen}
+            setIsModalOpen={setIsOptionListModalOpen}
+          />
+        );
+        break;
+      case 3:
+        modal = (
+          <DeleteFolderModal
+            folder={selectedFolder}
+            isModalOpen={isOptionListModalOpen}
+            setIsModalOpen={setIsOptionListModalOpen}
+          />
+        );
+        break;
+      default:
+        modal = null;
+        break;
+    }
+    return modal;
+  };
```
여기도 modalId를 조금 더 구체적으로 작성해주세요!

SeolJaeHyeok commented on src/components/Main/SortingSection.js:
```diff
@@ -65,27 +137,43 @@ function SortingSection({ selectedFolder, setSelectedFolder }) {
-                  className={selectedFolder.id === folder.id ? selectedButtonStyle : ''}
-                  onClick={() => handleButtonClick(folder.id)}
+                  className={
+                    selectedFolder.id === folder.id
+                      ? selectedButtonClasses
+                      : hoveredFolder?.id === folder.id
+                        ? hoveredButtonClasses
+                        : defaultButtonClasses
```
삼항 연산자가 중첩되면 가독성이 매우 나빠지게 됩니다 ㅜㅜ 따로 변수로 분리할 수 있는지 확인하고 분리할 수 있다면 분리해주세요!

SeolJaeHyeok commented on src/components/Main/SortingSection.js:
```diff
@@ -65,27 +137,43 @@ function SortingSection({ selectedFolder, setSelectedFolder }) {
+                        : defaultButtonClasses
+                  }
+                  onClick={() => handleSortingButtonClick(folder.id)}
+                  onMouseEnter={() => handleSortingButtonMouseEnter(folder.id)}
+                  onMouseLeave={handleSortingButtonMouseLeave}
```
여기도 불필요하게 리렌더링이 많이 일어나겠네요

SeolJaeHyeok reviewed (commented):
9주차 미션 고생많으셨습니다 진호님👍 역시 잘해주셨네요!

모달은 앞으로 개발하면서 정말 많이 만들고 사용하게 되는 컴포넌트 중 하나가 될텐데요.
관리하는 방법 또한 정말 다양하게 존재합니다!

지금처럼 컴포넌트 안에서 관리해도 되지만 멘토링때도 그렇고 아래 리뷰에서도 그렇고 컴포넌트 내부에서 관리하게 되면 나중에 모달이 늘어나게 되는 경우 그에 맞게 컴포넌트 내의 코드도 커지게 되는 일이 발생할 수 있어요!

공통으로 처리할 수 있는 방법에 대해서 조금 더 고민해보시면 더 좋은 코드가 나올거라고 믿어 의심치 않습니다..!

jinhok96 committed:
chore: 메뉴 아이디 명시적으로 수정
	modified:   src/components/Common/SelectMenu.js

jinhok96 committed:
chore: 필요없는 stopPropagation 제거
	modified:   src/components/Main/Card.js

jinhok96 committed:
chore: 불필요한 렌더링 발생시키는 로직 제거
- css hover 스타일로 대체
	modified:   src/components/Common/SelectMenu.js
	modified:   src/components/Common/SelectMenu.module.css

jinhok96 committed:
chore: 모달 id 명시적으로 수정
	modified:   src/components/Main/SortingSection.js

jinhok96 committed:
chore: SortingSection 버튼 스타일 로직 개선
- hover 스타일을 css로 정의하여 불필요한 렌더링 제거
	modified:   src/components/Main/SortingSection.js
	modified:   src/components/Main/SortingSection.module.css

jinhok96 committed:
chore: ModalContext, ModalProvider 생성
	new file:   src/context/ModalContext.js
	new file:   src/context/ModalProvider.js

jinhok96 committed:
chore: index.js에 ModalProvider 추가
	modified:   src/index.js

jinhok96 committed:
chore: 주석 추가
	modified:   src/index.js

jinhok96 committed:
chore: useModal 추가
	new file:   src/hooks/useModal.js

jinhok96 committed:
chore: ModalProvider의 close 로직 개선
- 닫고자 하는 모달과 현재 열려있다고 저장된 모달이 동일할 경우에 close 실행
	modified:   src/context/ModalProvider.js
	modified:   src/hooks/useModal.js

jinhok96 committed:
chore: index.js에 모달 렌더링 위치 지정
	modified:   src/components/Modal/ModalContainer.js
	modified:   src/index.js

jinhok96 committed:
chore: react-modal 설치
	modified:   package-lock.json
	modified:   package.json

jinhok96 committed:
chore: ReactModal 추가
	modified:   src/index.js

jinhok96 committed:
chore: ModalContainer에 ReactModal 적용
	modified:   src/components/Modal/ModalContainer.js
	modified:   src/index.js

jinhok96 committed:
chore: props 변경, 호버 로직 수정
	modified:   src/components/Modal/AddFolderModal.js
	modified:   src/components/Modal/AddToFolderModal.js
	modified:   src/components/Modal/AddToFolderModal.module.css
	modified:   src/components/Modal/DeleteFolderModal.js
	modified:   src/components/Modal/DeleteLinkModal.js
	modified:   src/components/Modal/EditFolderNameModal.js
	modified:   src/components/Modal/ShareModal.js

jinhok96 committed:
feat: 모달 관련 여러 변경사항 적용
- ReactModal 적용
- Context API 적용
- useModal 커스텀 훅 적용
- 모달 호출 기능을 Modal.js에 저장
- 여러가지 모달 및 스타일 관련 로직 개선
	modified:   package-lock.json
	modified:   package.json
	modified:   public/index.html
	modified:   src/components/Common/Button.js
	modified:   src/components/Common/SelectMenu.js
	modified:   src/components/Header/AddLinkBar.js
	modified:   src/components/Main/SortingSection.js
	modified:   src/components/Modal/AddFolderModal.js
	modified:   src/components/Modal/AddToFolderModal.js
	modified:   src/components/Modal/AddToFolderModal.module.css
	new file:   src/components/Modal/Modal.js
	modified:   src/components/Modal/ModalContainer.js
	modified:   src/components/Modal/ShareModal.js
	modified:   src/context/ModalProvider.js
	modified:   src/index.js

jinhok96 committed:
chore: 필요없는 stopPropagation 제거
	modified:   src/components/Modal/AddFolderModal.js
	modified:   src/components/Modal/AddToFolderModal.js
	modified:   src/components/Modal/DeleteLinkModal.js
	modified:   src/components/Modal/ModalContainer.js

jinhok96 committed:
chore: 모달이 렌더링되는 위치를 #modal-root로 지정
	modified:   src/components/Modal/ModalContainer.js
	modified:   src/index.js

jinhok96 committed:
refactor: Modal.js 경로 변경
	modified:   src/components/Common/SelectMenu.js
	modified:   src/components/Header/AddLinkBar.js
	modified:   src/components/Main/SortingSection.js
	renamed:    src/components/Modal/Modal.js -> src/context/Modal.js
	modified:   src/context/ModalProvider.js

jinhok96 committed:
chore: 디버깅용 일부 콘솔 출력문을 제거하고 주석 추가
	modified:   src/components/Modal/AddFolderModal.js
	modified:   src/context/ModalContext.js
	modified:   src/context/ModalProvider.js
	modified:   src/index.js

jinhok96 committed:
chore: input의 value가 빈 값일 경우 클릭이 되지 않도록 수정
chore: 디버깅용 콘솔 출력문 수정
	modified:   src/components/Modal/EditFolderNameModal.js

jinhok96 committed:
feat: StarButton onClick 이벤트 추가, 클릭 시 스타일 변경
	modified:   src/assets/images/star.svg
	modified:   src/components/Common/StarButton.js
	modified:   src/components/Main/Card.js

jinhok96 committed:
chore: Loading... 위치 변경
	modified:   src/components/Header/Gnb.js

SeolJaeHyeok commented on src/context/Modal.js:
```diff
@@ -0,0 +1,42 @@
+};
+
+// 다른 곳에서 불러올 모달 컴포넌트
+function Modal() {
+  const openedModal = useContext(ModalStateContext);
+  const { close } = useContext(ModalDispatchContext);
+
+  if (!openedModal) {
+    return null;
+  }
+
+  const { ModalComponent, propList } = openedModal;
+  const { onSubmit, ...restPropList } = propList;
+
+  const onClose = () => {
+    close(ModalComponent);
+  };
+
+  // onSubmit -> 동기/비동기 여부를 모르기 때문에 비동기로 처리
+  const handleSubmit = async () => {
+    await onSubmit();
+    onClose();
+  };
+
+  return <ModalComponent {...restPropList} onSubmit={handleSubmit} onClose={onClose} />;
+}
+
+export { modalList, Modal };
```
```suggestion
function Modal() {
  const openedModal = useContext(ModalStateContext);
  const { close } = useContext(ModalDispatchContext);

  if (!openedModal) {
    return null;
  }

  const { ModalComponent, propList } = openedModal;
  const props = propList;

  const onClose = () => {
    close(ModalComponent);
  };

  // onSubmit -> 동기/비동기 여부를 모르기 때문에 비동기로 처리
  const handleSubmit = async () => {
    if (props && props.onSubmit) {
      await props.onSubmit();
    } 
    onClose();
  };

  return <ModalComponent {...props} onSubmit={handleSubmit} onClose={onClose} />;
}

export { modalList, Modal };
```
너무 잘 만들어주셨는데, 지금 인자로 주입받은 propsList에서 onSubmit을 구조분해할당하고 있는 부분에서 `onSubmit` 함수가 없는 경우에 대한 예외처리가 필요할거 같아요!

TypeScript를 쓰게 된다면 onSubmit을 필수값으로 지정하는 방식 등을 통해서(물론 여기서도 예외 처리가 필요합니다만) 어느정도 해결은 가능하지만 지금은 JavaScript 환경이고 아시다시피 JavaScript를 인자 주입과 관련해서 제약이 없기 때문에 꼼꼼한 예외처리가 필요합니다!

SeolJaeHyeok commented on src/index.js:
```diff
@@ -10,9 +13,21 @@ import 'assets/styles/utility.css';
+
 const root = ReactDOM.createRoot(document.getElementById('root'));
+
+// React.StrictMode: 개발 단계에서 안전하지 않은 생명주기, 레거시 API 사용, 예상치 못한 부작용 등을 개발 단계에서 미리 발견하고 수정
+// React.StrictMode -> render(), shouldComponentUpdate(), setState, ... 등 여러 메서드가 두 번 호출됨
+// React.StrictMode -> 모달이 두 번 호출 -> 두 번 렌더링되어 'React-Modal: Cannot register modal instance that's already open' 오류가 발생
+// React.StrictMode는 프로덕션 빌드에서는 작동하지 않음
+// ModalProvider로 App을 감싸면 App의 모든 요소에서 ModalProvider에 접근 가능
```
Strict Mode 정리 아주 좋습니다! 👍
이 부분은 팀프로젝트하면서 팀원들한테도 말해줘도 될 거 같아요!

SeolJaeHyeok reviewed (commented):
너무 잘해주셔서 깜짝 놀랐습니다 진짜 ㅎㅎ 
큰 틀은 이야기 드릴게 없을 정도여서 디테일한 부분에 대해서 멘트 하나 남겼으니 확인 해주세요~!

jinhok96 committed:
chore: propList와 onSubmit이 없는 경우 예외 처리
	modified:   src/context/Modal.js
